[mlir] Make the split markers of splitAndProcessBuffer configurable.

ingomueller-net opened the pull request:
This allows to define custom splitters, which is interesting for non-MLIR inputs and outputs to `mlir-translate`. For example, one may use `; -----` as a splitter of `.ll` files. The splitters are now passed as arguments into `splitAndProcessBuffer`, the input splitter defaulting to the previous default (`// -----`) and the output splitter defaulting to the empty string, which also corresponds to the previous default. The behavior of the input split marker should not change at all; however, outputs now have one new line *more* than before if there is no splitter (old: `insertMarkerInOutput = false`, new: `outputSplitMarker = ""`) and one new line *less* if there is one. The value of the input splitter is exposed as a command line options of `mlir-translate` and other tools as an optional value to the previously existing flag `-split-input-file`, which defaults to the default splitter if not specified; the value of the output splitter is exposed with the new `-output-split-marker`, which default to the empty string in `mlir-translate` and the default splitter in the other tools. In short, the previous usage or omission of the flags should result in previous behavior (modulo the new lines mentioned before).

ingomueller-net committed:
Make filt split marker ('// -----') a global constant.

This allows all tools to use that constant defined in `ToolUtilities.h`,
instead of copying that string.

ingomueller-net committed:
Make the split markers of splitAndProcessBuffer configurable.

joker-eph commented on mlir/lib/Tools/mlir-translate/MlirTranslateMain.cpp:
```diff
@@ -80,6 +80,14 @@ LogicalResult mlir::mlirTranslateMain(int argc, char **argv,
 
+  static llvm::cl::opt<std::string> inputSplitMarker(
+      "input-split-marker", llvm::cl::desc("Split marker to use for the input"),
+      llvm::cl::init(kDefaultSplitMarker));
```
Instead of a separate flag, can we just change `splitInputFile` to be a string?

ingomueller-net committed:
Merge `-input-split-marker` into `-split-input-file`.

In this new version, specifying `-split-input-file` without value uses
the default marker (`// -----`) implicitly, which splits the input on
that marker, whereas not specifying that flag uses the empty marker,
which does not split the input. This makes all previous usages of the
flag work as before (modulo the subtle new-line changes introduced in
the previous commit). A custom splitter can be used by providing a value
to the same flag (`-split-input-file="; -----"`).

ingomueller-net commented on mlir/lib/Tools/mlir-pdll-lsp-server/PDLLServer.cpp:
```diff
@@ -1621,7 +1622,8 @@ PDLTextFile::getPDLLViewOutput(lsp::PDLLViewOutputKind kind) {
         [&](PDLTextFileChunk &chunk) {
           chunk.document.getPDLLViewOutput(outputOS, kind);
         },
-        [&] { outputOS << "\n// -----\n\n"; });
```
Note that I am not sure whether this and the other occurrence of the constant in the same file below *have* to be the same as in `mlir-opt` and `mlir-translate` or whether it this constant is local to the PDLL server. In the latter case, it might be better not to use the constant from `ToolUtilities.h`. What do you think?

joker-eph replied:
This is fine..

joker-eph commented on mlir/lib/Support/ToolUtilities.cpp:
```diff
@@ -89,7 +88,7 @@ mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer> originalBuffer,
   };
   llvm::interleave(sourceBuffers, os, interleaveFn,
-                   insertMarkerInOutput ? "\n// -----\n" : "");
+                   (llvm::Twine(outputSplitMarker) + "\n").str());
```
This inserts a `\n` even if `outputSplitMarker` is empty, it's a change from the previous code, but I don't know if it matters!

ingomueller-net replied:
I thought a bit about this. I think the trade-off is between exact preservation of previous behavior or consistency with itself. Indeed, this is a change (as described in the commit message) but the tests in this repo still all pass. I'd argue that the current formulation is consistent in that it adds the output split marker, which might be empty, plus a new line.

It's true that this makes it impossible to specify **no** separation. But if we make a special case for the empty separator, then we still introduce a implicit new line after all other separators. And I don't see a way to change that: not only would it be cumbersome to specify a new line in a command line argument, but it is outright impossible to do so at the end of a command line value because LLVM's parsing library skips over all trailing whitespace.

Anyways, I've added a sentence about the new lines to the doc. If anyone has arguments or strong feelings against the current behavior, I am happy to still change it.

joker-eph commented:
You need to look into the CI failures I think

ingomueller-net committed:
Fix CI: Add full CL parsing logic to mlir-pdll.

I have forgotten to add the extended logic from the previous commit to
that tool. This commit fixes that and adds a corresponding test.

ingomueller-net commented on mlir/include/mlir/Support/ToolUtilities.h:
```diff
@@ -27,20 +29,22 @@ struct LogicalResult;
-                      bool enableSplitting = true,
-                      bool insertMarkerInOutput = false);
+                      llvm::StringRef inputSplitMarker = kDefaultSplitMarker,
+                      llvm::StringRef outputSplitMarker = kDefaultSplitMarker);
```
No. Good catch! Fixed with latest commit.

ingomueller-net committed:
Make default for outputSplitMarker match previous default behavior.

ingomueller-net committed:
Document insertion of the new line character inserted by splitAndProcessBuffer.

ingomueller-net reviewed (commented):
Thanks for the review! CI and accidental change of behavior are now fixed. I also left some rationale about the new line character.

ingomueller-net committed:
Rename test/mlir-opt/{nearmiss,split-markers}.mlir.

akuegel commented on mlir/include/mlir/Tools/mlir-opt/MlirOptMain.h:
```diff
@@ -136,13 +137,24 @@ class MlirOptMainConfig {
+    splitInputFileFlag = std::move(splitMarker);
+    return *this;
+  }
+  bool shouldSplitInputFile() const { return splitInputFileFlag.empty(); }
```
I would expect that this should return true if splitInputFileFlag is not empty? Otherwise we will try to use empty string as split marker?
I think that IREE makes use of this function, otherwise it seems it is currently untested.

ingomueller-net replied:
Yeah, I think you are right. It should be `!splitInputFileFlag.empty()`.
